File: c/tree/avl.c
#include <stdio.h>
#include <stdlib.h>

typedef struct node{
    int key;
    struct node*left,*right;
    int height;
}n;

int getheight(n* node){
    return node?node->height:-1;
}

int max(int a,int b){
    return (a>b)?a:b;
}

n* createnode(int key){
    n* newnode=(n*)malloc(sizeof(n));
    newnode->key=key;
    newnode->left=newnode->right=NULL;
    newnode->height=0;
    return newnode;
}

void rotateright(n** root){
    n* node=*root;
    n* ptr=node->left;
    n* ptr2=ptr->right;
    ptr->right=node;
    node->left=ptr2;
    node->height=max(getheight(node->left),getheight(node->right))+1;
    ptr->height=max(getheight(ptr->left),getheight(ptr->right))+1;
    *root= ptr;
}

void rotateleft(n** root){
    n* node=*root;
    n* ptr=node->right;
    n* ptr2=ptr->left;
    ptr->left=node;
    node->right=ptr2;
    node->height=max(getheight(node->left),getheight(node->right))+1;
    ptr->height=max(getheight(ptr->left),getheight(ptr->right))+1;
    *root= ptr;
}
int getbalance(n* node){
    return node?getheight(node->left)-getheight(node->right):0;
}

void insert(n**root,int key){
    if(!(*root)){
        *root=createnode(key);
        return;
    }
    if(key<(*root)->key){
        insert(&(*root)->left,key);
    }else if(key>(*root)->key){
        insert(&(*root)->right,key);
    }else{
        return;
    }

    (*root)->height=max(getheight((*root)->left),getheight((*root)->right))+1;
    int balance=getbalance(*root);
    if(balance>1 && key<(*root)->left->key){
        rotateright(root);
    }
    if(balance<-1 && key> (*root)->right->key){
        rotateleft(root);
    }
    if(balance>1 && key> (*root)->left->key){
        rotateleft(&(*root)->left);
        rotateright(root);
    }
    if(balance<-1 && key<(*root)->right->key){
        rotateright(&(*root)->right);
        rotateleft(root);
    }
    
}

void inorder(n*root) {
    if (root) {
        inorder(root->left);
        printf("%d ", root->key);
        inorder(root->right);
    }
}

int main(){
    n* root=NULL;
    insert(&root,9);
    insert(&root,15);
    insert(&root,20);
    insert(&root,8);
    insert(&root,7);
    insert(&root,13);
    insert(&root,10);
    printf("inorder:");
    inorder(root);
}

File: python/tree/avl.py
class Node:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None
        self.height = 0

class AVLTree:
    def __init__(self):
        self.root = None

    def get_height(self, node):
        return node.height if node else -1

    def max(self, a, b):
        return a if a > b else b

    def rotate_right(self, root):
        node = root
        ptr = node.left
        ptr2 = ptr.right

        ptr.right = node
        node.left = ptr2

        node.height = self.max(self.get_height(node.left), self.get_height(node.right)) + 1
        ptr.height = self.max(self.get_height(ptr.left), self.get_height(ptr.right)) + 1

        return ptr

    def rotate_left(self, root):
        node = root
        ptr = node.right
        ptr2 = ptr.left

        ptr.left = node
        node.right = ptr2

        node.height = self.max(self.get_height(node.left), self.get_height(node.right)) + 1
        ptr.height = self.max(self.get_height(ptr.left), self.get_height(ptr.right)) + 1

        return ptr

    def get_balance(self, node):
        return self.get_height(node.left) - self.get_height(node.right) if node else 0

    def insert(self, root, key):
        if not root:
            return Node(key)

        if key < root.key:
            root.left = self.insert(root.left, key)
        elif key > root.key:
            root.right = self.insert(root.right, key)
        else:
            return root

        root.height = self.max(self.get_height(root.left), self.get_height(root.right)) + 1
        balance = self.get_balance(root)

        # Left-Left case
        if balance > 1 and key < root.left.key:
            return self.rotate_right(root)

        # Right-Right case
        if balance < -1 and key > root.right.key:
            return self.rotate_left(root)

        # Left-Right case
        if balance > 1 and key > root.left.key:
            root.left = self.rotate_left(root.left)
            return self.rotate_right(root)

        # Right-Left case
        if balance < -1 and key < root.right.key:
            root.right = self.rotate_right(root.right)
            return self.rotate_left(root)

        return root

    def inorder(self, root):
        if root:
            self.inorder(root.left)
            print(root.key, end=" ")
            self.inorder(root.right)

    def insert_key(self, key):
        self.root = self.insert(self.root, key)

# Usage
if __name__ == "__main__":
    tree = AVLTree()
    keys = [9, 15, 20, 8, 7, 13, 10]

    for key in keys:
        tree.insert_key(key)

    print("Inorder traversal:", end=" ")
    tree.inorder(tree.root)
